### ISSUE TITLE
Wrong CPU multipliers on AMD APU

### ISSUE THREAD

**Issue by X0rg** [anrieff/libcpuid#52]
As reported in X0rg/CPU-X#31, CPU multipliers are wrong with some AMD APUs.
The minimum and maximum multipliers are wrong, and the bus clock is wrong too.

**Possible causes:**
* Wrong P-state read
* Wrong formula for multiplier computation

**What is needed to fix it:**
* More tests

There is a standalone source code on my [Gist](https://gist.githubusercontent.com/X0rg/150f8effb64d70db3544d228f2348a9e/raw/2f08f0f8bd1e795d0b78b9a698aa31d4486b0dc7/apu_msr.c) which can be used to help to debug. Build and run instructions are on the top of file.
I'm looking for testers.

**Comment by anrieff** [comment]
Hi, I poked a friend of mine to test your gist, he has a C50/C60, so not exactly the same, but might be still helpful. Will keep you posted on the results.

**Comment by anrieff** [comment]
Here's one from **E-350** (family 0x0F, ext_family 0x14, ext_model 0x01):

```
Maximum non-turbo speed: 1596
Highest-performance non-boosted P-state: 0
Addr: C0010064 FID: 16 DID: 0 Mult: 32 (err=0)
Addr: C0010065 FID: 18 DID: 0 Mult: 34 (err=0)
Addr: C0010066 FID: 48 DID: 0 Mult: 64 (err=0)
Addr: C0010067 FID: 0 DID: 0 Mult: 16 (err=0)
Addr: C0010068 FID: 0 DID: 0 Mult: 16 (err=0)
Addr: C0010069 FID: 0 DID: 0 Mult: 16 (err=0)
Addr: C001006A FID: 0 DID: 0 Mult: 16 (err=0)
Addr: C001006B FID: 0 DID: 0 Mult: 16 (err=0)
```

**Comment by X0rg** [comment]
Thanks a lot!
The minimum multiplier for this CPU is 8, and the maximum is 16, according to [this screenshot](http://cdn.cpu-world.com/Images/uploaded/0000/69/L_00006912.jpg).
Multipliers are computed on [line 584](https://github.com/anrieff/libcpuid/blob/master/libcpuid/rdmsr.c#L584) with:
`multiplier = (uint64_t) (((CpuFid + 0x10) / (1ull << CpuDid)) / factor);`
A factor of 2.0 instead 1.0 should fix this issue for this CPU.
The bus clock is `1596/16 = 99,75MHz`, that seems good.

I'll try to find more people with an APU, because I don't understand why the bus clock is 549MHz on A10-5800K...

**Comment by anrieff** [comment]
Another data point from a **C-60** (family 0x0F, ext_family 0x14, ext_model 0x02).

```
Maximum non-turbo speed: 998
Highest-performance non-boosted P-state: 0
Addr: C0010064    FID: 32    DID: 0    Mult: 48 (err=0)
Addr: C0010065    FID: 48    DID: 0    Mult: 64 (err=0)
Addr: C0010066    FID: 0    DID: 1    Mult: 8 (err=0)
Addr: C0010067    FID: 18    DID: 1    Mult: 17 (err=0)
Addr: C0010068    FID: 0    DID: 0    Mult: 16 (err=0)
Addr: C0010069    FID: 0    DID: 0    Mult: 16 (err=0)
Addr: C001006A    FID: 0    DID: 0    Mult: 16 (err=0)
Addr: C001006B    FID: 0    DID: 0    Mult: 16 (err=0)
```

**Comment by X0rg** [comment]
Ok, thank you.
The C-60 should have 3 mutlipliers:
- 8, the minimum.
- 10, the maximum.
- 13.3, the boosted state.

It seems CPU multiplier calculation is not the same for all families... For [14h family](http://support.amd.com/TechDocs/43170_14h_Mod_00h-0Fh_BKDG.pdf), `CpuDID = CpuDidMSD + (CpuDidLSD * 0.25) + 1`.

**Comment by anrieff** [comment]
OK, just making sure we're on the same boat: I believe you should be fixing this issue when you find what the correct multiplier calculation logic is. Let me know if I can be of any further help.

**Comment by X0rg** [comment]
Yes, I'll not fix it before finding multiplier calculation. It is very different than Intel logic.
These tests are very useful for me, it shows multiplier calculation is not the same for all, so I need to read more about it.
That's all for tests, no more are needed. I'll fix it by myself, read the AMD's BKDG is a bit boring, but I have free time.

I'm sorry for this bug.

**Comment by X0rg** [comment]
I share more tests, given by someone:

**- AMD E-350**
```
Maximum non-turbo speed: 1599
Highest-performance non-boosted P-state: 0 (addr: C0010064)
P-state for minimum CPU multiplier: 2
Addr: C0010064   FID: 16   DID: 0   Mult: 32 (err=0)
Addr: C0010065   FID: 18   DID: 0   Mult: 34 (err=0)
Addr: C0010066   FID: 48   DID: 0   Mult: 64 (err=0)
Addr: C0010067   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C0010068   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C0010069   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C001006A   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C001006B   FID: 0   DID: 0   Mult: 16 (err=0)
```
==> Wrong: multipliers should be [[8-16]](http://cdn.cpu-world.com/Images/uploaded/0000/69/L_00006912.jpg).


**- AMD Sempron 3850**
```
Maximum non-turbo speed: 1300
Highest-performance non-boosted P-state: 0 (addr: C0010064)
P-state for minimum CPU multiplier: 3
Addr: C0010064   FID: 10   DID: 1   Mult: 13 (err=0)
Addr: C0010065   FID: 8   DID: 1   Mult: 12 (err=0)
Addr: C0010066   FID: 4   DID: 1   Mult: 10 (err=0)
Addr: C0010067   FID: 0   DID: 1   Mult: 8 (err=0)
Addr: C0010068   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C0010069   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C001006A   FID: 0   DID: 0   Mult: 16 (err=0)
Addr: C001006B   FID: 0   DID: 0   Mult: 16 (err=0)
```
==> Good: multipliers are [[8-13]](http://www.xbitlabs.com/images/cpu/athlon-5350-sempron-3850/cpuz-3850.png).

**Comment by X0rg** [comment]
A (temporary) patch is pending on [my Gist](https://gist.githubusercontent.com/X0rg/150f8effb64d70db3544d228f2348a9e/raw/b1bedee6898960bc9ccecfa63e2465b1bdc5c28c/apu_msr.c).
New output for **Sempron 3850**:
```
Maximum non-turbo speed: 1300
Highest-performance non-boosted P-state: 0 (addr: 0xC0010064)
P-state for minimum CPU multiplier: 3 (addr: 0xC0010067)
Dump for P0 (0xC0010064): 0x800001240000504A
Dump for P1 (0xC0010065): 0x8000011F00005448
Dump for P2 (0xC0010066): 0x8000011A00006444
Dump for P3 (0xC0010067): 0x800002DE00407440
Dump for P4 (0xC0010068): 0x1FE00
Dump for P5 (0xC0010069): 0x1FE00
Dump for P6 (0xC001006A): 0x1FE00
Dump for P7 (0xC001006B): 0x1FE00

Values provided by libcpuid:
Multipliers: [4.00-6.00]        Bus clock: 216.50 MHz

Values provided by patch:
[K10] Multipliers: [8-13]       Bus clock: 100.00 MHz (x13)
[K11] Multipliers: [4-9]        Bus clock: 144.44 MHz (x9)
[K12] Multipliers: [20-20]      Bus clock: 65.00 MHz (x20)
[K14] Multipliers: [0-0]        Bus clock: inf MHz (x0)
[K15] Multipliers: [8-13]       Bus clock: 100.00 MHz (x13)
[K16] Multipliers: [8-13]       Bus clock: 100.00 MHz (x13)
```
It is better. But I have some questions before writing the final patch:

1. On 14h family, CPU COF is computed with `D18F3xD4[MainPllOpFreqId]`, but it is a PCI address, not a MSR register. I don't know if I can retrieve this value. An idea how to do this?

2. Does anybody know if AMD CPU family (like 10h, 16h...) correspond to CPUID Extended Family?

**Comment by X0rg** [comment]
For 1., I have found a possible workaround, see [this Gist](https://gist.githubusercontent.com/X0rg/150f8effb64d70db3544d228f2348a9e/raw/647672ad303d9f03106e6d4f3607c0440089bd94/apu_msr.c). It seems `MainPllOpFreqId` is equal to `(cpu_clock / 100 + 0x10)`, because `PLL COF = 100 MHz * (D18F3xD4[MainPllOpFreqId] + 10h`.

On a **E-350** (14h family) without this workaround:
```
Maximum non-turbo speed: 1599
Highest-performance non-boosted P-state: 0 (addr: 0xC0010064)
P-state for minimum CPU multiplier: 2 (addr: 0xC0010066)
Dump for P0 (0xC0010064): 0x8000012600002810
Dump for P1 (0xC0010065): 0x8000011A00003612
Dump for P2 (0xC0010066): 0x8000028C00005E30
Dump for P3 (0xC0010067): 0x0
Dump for P4 (0xC0010068): 0x0
Dump for P5 (0xC0010069): 0x0
Dump for P6 (0xC001006A): 0x0
Dump for P7 (0xC001006B): 0x0

Values provided by libcpuid:
Multipliers: [64.00-32.00]   Bus clock: 49.96 MHz

Values provided by patch:
[K10] Multipliers: [64-32]   Bus clock: 49.97 MHz (x32)
[K11] Multipliers: [56-24]   Bus clock: 66.62 MHz (x24)
[K12] Multipliers: [19-17]   Bus clock: 94.06 MHz (x17)
[K14] Multipliers: [0-0]   Bus clock: inf MHz (x0)
[K15] Multipliers: [64-32]   Bus clock: 49.97 MHz (x32)
[K16] Multipliers: [64-32]   Bus clock: 49.97 MHz (x32)
```
For P0 (0x8000012600002810), CpuDidMSD = 1 and CpuDidLSD = 0. With formula, ((1599+5)/100+16)/(1+0+1) = **16**.
For P2 (0x8000028C00005E30), CpuDidMSD = 3 and CpuDidLSD = 0. With formula,((1599+5)/100+16)/(3+0+1) = **8.**
(The `+5` is a fake round, due to 1599 should be 1600.)
So, multipliers are [8-16] with this workaround, and it is great.

**Comment by anrieff** [comment]
Hi @X0rg,

your progress is excellent. On your question 2), see cpuid_basic_identify() in cpuid_main.c:

		xfamily = (raw->basic_cpuid[1][0] >> 20) & 0xff;
		if (data->vendor == VENDOR_AMD && data->family < 0xf)
			data->ext_family = data->family;
		else
			data->ext_family = data->family + xfamily;

This code might be as well wrong, since it's been written a very long time ago.

I very much feel that the raw MSR data & expected decoded data should be included in the test cases. I'll think about it.

**Comment by phprus** [comment]
Hi, @anrieff,

Visual Studio and more compilers does not support binary literal in C. (0b0000 - 0b1000, https://github.com/anrieff/libcpuid/pull/56/files#diff-1e8f13971608fa43c3ac87209c916579R604).

**Comment by anrieff** [comment]
Thanks for the notice, will fix them shortly.

**Comment by X0rg** [comment]
Sorry for that, that's because AMD use binary literal in BKDG. Fixed in #57.
